### ISSUE TITLE
Error in connector: Error creating a database connection. (Operation timed out (mobc timeout))

### ISSUE THREAD

**Issue by nicosampler** [prisma/prisma#4280]
## Bug description
Sometimes, due to DB connections limit, the client fails with this error: 

```
Error in connector: Error creating a database connection. (Operation timed out (mobc timeout))
    at PrismaClientFetcher.request (/home/nicosampler/develop/loan-system/node_modules/@prisma/client/runtime/index.js:78786:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async eval (webpack-internal:///./graphql/migration/resolver.ts:139:17)
(node:106069) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 75)
```

my DB configuration is like so: `DATABASE_URL="mysql://someUser:somePass@host/db?connection_limit=10"`

**Note:** `connection_limit`

## How to reproduce

It happened to me when trying to create/update/delete several entities at the same time, for example:

```
const res = [largeArray]
res.forEach(async (client) => {
       const createdClient = await context.prisma.client.create({data: {client}})
})
```

## Expected behavior
The client should be able to recover from a timeout connection. 

## Environment & setup
- OS: Ubuntu
- Database: Mysql
- Node.js version: 14
- Prisma version: 2.11.0

**Comment by mayerlench** [comment]
Im having the same issue

`error PrismaClientUnknownRequestError2:
Invalid `prisma.item.create()` invocation:


  Error in connector: Error creating a database connection. (Operation timed out (mobc timeout))
    at PrismaClientFetcher.request (C:\Users\MayerLench\Documents\VSCode\Seasons\pca-price-comparison\node_modules\@prisma\client\runtime\index.js:78786:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  clientVersion: '2.11.0'
}`

**Comment by alfonmga** [comment]
Same issue here.

```json
{
  "data": {
    "message": "Error: Error creating a database connection. (Operation timed out (mobc timeout))",
    "reason": {
      "clientVersion": "2.11.0",
      "message": "Error creating a database connection. (Operation timed out (mobc timeout))",
      "stack": "Error: Error creating a database connection. (Operation timed out (mobc timeout))\n    at LineStream.<anonymous> (/var/task/node_modules/@prisma/client/runtime/index.js:27430:31)\n    at LineStream.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at LineStream.Readable.push (_stream_readable.js:212:10)\n    at LineStream.Transform.push (_stream_transform.js:152:32)\n    at LineStream._pushBuffer (/var/task/node_modules/@prisma/client/runtime/index.js:24491:19)\n    at LineStream._transform (/var/task/node_modules/@prisma/client/runtime/index.js:24485:10)\n    at LineStream.Transform._read (_stream_transform.js:191:10)\n    at LineStream.Transform._write (_stream_transform.js:179:12)"
    },
    "stack": "Runtime.UnhandledPromiseRejection: Error: Error creating a database connection. (Operation timed out (mobc timeout))\n    at process.<anonymous> (/var/runtime/index.js:35:15)\n    at process.emit (events.js:327:22)\n    at processPromiseRejections (internal/process/promises.js:209:33)\n    at processTicksAndRejections (internal/process/task_queues.js:98:32)"
  },
  "level": "error",
  "message": "Unhandled Promise Rejection \t",
  "timestamp": "2020-11-20T07:03:55.934Z"
}
```

BTW I'm using [Blitz.js](https://github.com/blitz-js/blitz).

**Comment by ajdam10** [comment]
I am having the same issue.

**Comment by assertnotnull** [comment]
We face this also at my work and we are using pgBouncer.
What would be really useful is a stacktrace when possible because it's very hard to find out where is the problem.

**Comment by alfonmga** [comment]
My comment in #2782

> I'm experiencing the same issue `Error creating a database connection. (Operation timed out (mobc timeout))` and I have been tweaking the `connect_timeout` and `socket_timeout` parameters in my connection string but looks like it has no effect even when I change them to _absurd_ values (9999 seconds): instead I get the timeout error right away without waiting for the defined timeout value. May this be a bug?
> 
> Here's the connection string I'm using on prod:
> 
> ```
> postgresql://fastmessage-io:<pw>@amgaventures-do-user-8339649-0.b.db.ondigitalocean.com:25061/fastmessage-io?sslmode=require&pgbouncer=true&connection_limit=1&connect_timeout=10&socket_timeout=10
> ```
> 
> This how Prisma is instanced (I'm using [Blitz.js](https://github.com/blitz-js/blitz)):
> 
> ```ts
> import { PrismaClient } from "@prisma/client"
> export * from "@prisma/client"
> 
> let prisma: PrismaClient
> 
> if (process.env.NODE_ENV === "production") {
>   prisma = new PrismaClient()
> } else {
>   // Ensure the prisma instance is re-used during hot-reloading
>   // Otherwise, a new client will be created on every reload
>   globalThis["prisma"] = globalThis["prisma"] || new PrismaClient()
>   prisma = globalThis["prisma"]
> }
> 
> export default prisma
> ```
> 
> Prisma client version: `2.11.0`

I think this timeout issue could be solved easily by tweaking `connect_timeout` and `socket_timeout` connection string params but as I commented in that comment they don't work as expected. 

Could you guys give it a try? all you need to reproduce this issue is to create a small pool size (1-2) and fire some concurrent Prisma queries. If you get a timeout error immediately without wait for the defined timeout then there's some bug and we should open another issue to tackle it.

**Comment by Stevenje** [comment]
+1
```
PrismaClientUnknownRequestError2: 
Invalid `prisma.campaign.upsert()` invocation:


  Error in connector: Error creating a database connection. (Operation timed out (mobc timeout))
    at PrismaClientFetcher.request (/Users/steven/Documents/boilerplate/versiontwo/node_modules/@prisma/client/runtime/index.js:79359:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:93:5)
    at /Users/steven/Documents/boilerplate/versiontwo/packages/amazon/lib/amazon.ts:483:42 {
  clientVersion: '2.12.1'
}
```

**Comment by assertnotnull** [comment]
This error shows with or without pgBouncer or cloudsql-proxy.

**Comment by pantharshit00** [comment]
I can reproduce this with a heroku database with connection limit of 10 and this simple script.

![image](https://user-images.githubusercontent.com/22195362/101621421-bb9c5b80-3a3b-11eb-860c-3f404cee29f9.png)

All timed out calls doesn't recover. Connections in the heroku reaches max of 11 then it starts to give timeouts.

![image](https://user-images.githubusercontent.com/22195362/101622491-1e422700-3a3d-11eb-8735-b5853b19990b.png)



Reproduction: https://github.com/harshit-test-org/prisma-issue-4280

**Comment by MichalLytek** [comment]
@pantharshit00 I can reproduce that also for the `Can't reach database server` error, when the docker DB container is not started.

I found it as a regression because some time ago (a few weeks or months ago) it worked like a charm - I could start the db when the server was running and it could reconnect itself when e.g. the `findMany` was called (implicit connect).

**Comment by s-kris** [comment]
+1 
I can confirm the same issue on digital ocean 1GB RAM Postgres12 instance without pgbouncer.

**Comment by s-kris** [comment]
The same bug is occurring even after using custom-server for nextjs. (prisma instance is initialised once at start of the server and passed down the routes through req object).

**Comment by assertnotnull** [comment]
I think this problem is with the query engine https://www.prisma.io/docs/concepts/components/prisma-client/query-engine#the-query-engine-at-runtime we are facing a lot of stability problems with the query engine: timeouts, OOM..

**Comment by timsuchanek** [comment]
Thanks, everyone for your comments! We are looking into it right now and discuss solutions internally. We will definitely make sure to give you a solution for this.

**Comment by timsuchanek** [comment]
The first simple thing to try out, which we just talked about in the team:
Increasing the `connection_limit`. Can you please try out to increase it and see, if you are still getting the timeouts? Because of how Prisma is doing write operations, the connection pool can get saturated quite quickly. If you're then trying to do any other query, it can't even be performed, because the Pool is full. Having a bigger pool size is probably the most effective way to mitigate that.

You can check out [here](https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool) how to increase the `connection_limit`.

**Comment by nicosampler** [comment]
@timsuchanek I have tried previously that with the same result. When I'm running a big amount of operations sooner than later I ended with the same error.

**Comment by timsuchanek** [comment]
Thanks @nicosampler. Could you try a much higher value like 50 or 100?

**Comment by nicosampler** [comment]
> Thanks @nicosampler. Could you try a much higher value like 50 or 100?

I could try locally, but it won't be the case for my "production env". my database plan provides me only 10. 
So running an update in 50/60 records at the same time will produce this problem.

**Comment by JonathanCallewaert** [comment]
I am using the last version of prisma (2.15.0). It's throwing the new error (https://github.com/prisma/prisma-engines/pull/1515). But the error message doesn't make sense. I have 150 connections as limit, 10 connections are used but it still fires an error: 
```
Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 150, amount of connections currently in use: 10.
```

**Comment by seanking11** [comment]
I am also using the latest version of `@prisma/client` and `@prisma/cli` of `2.15.0` and am seeing that I'm getting the same error above, except it mentions that there are **0** connections in use.

I am deploying to Firebase cloud functions, running on Node 12.

```
Error: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 10, amount of connections currently in use: 0.
```

This is how I instantiate my PrismaClient

```
const Prisma = require('@prisma/client')

let prisma

if (process.env.NODE_ENV === 'production') {
    prisma = new Prisma.PrismaClient()
} else {
    if (!global.prisma) {
        global.prisma = new Prisma.PrismaClient()
    }

    prisma = global.prisma
}

module.exports = prisma
```

**Comment by jan-wilhelm** [comment]
the exact same thing that @seanking11 is seeing is also happening to me. On the database side (DigitalOcean Postgres + pgBouncer with 15 connections allowed) it only shows one connection utilized. The Prisma error also says `amount of connections currently in use: 0.`

**Comment by janpio** [comment]
We are looking into this.

What I can already say: 
The "in use" number is wrong and does not show what we expected it would show 💩. Sorry about that @JonathanCallewaert, @seanking11 and @jan-wilhelm. You can ignore it and anything it seems to be telling you. We will improve this 🙇 

Now to figure out what is going on:
- Is it possible something is creating _lot_ of queries in a short timeframe that would be possibly executed in parallel?
- Can you describe what is happening in your application?
- Can you try adding `&pool_timeout=0` to your connection string? This will try to totally disable this timeout and just let queries wait in the queue _forever_ until they can be run.

If you prefer sync chat, I am @janpio on slack.prisma.io.

**Comment by JonathanCallewaert** [comment]
@janpio thank you for your response!

I'm migrating data from an old environment to a new one. To improve speed, I spin up some child process (16) which runs some queries. All the child processes are using the same Prisma instance. However, it also occurs if I use a new PrismaClient per child process.

So I think the reason that Prisma triggers this is because of a lot of queries running in parallel as you mentioned. 

Adding the `&pool_timeout=0` doesn't seem to fix the issue :( . You mean to add it like this : `...m:5432/postgres?pool_timeout=0` , right?

**Comment by janpio** [comment]
> Adding the `&pool_timeout=0` doesn't seem to fix the issue :( . You mean to add it like this : `...m:5432/postgres?pool_timeout=0` , right?

Yes, if you have no other query parameters in your connection URL it would be `?pool_timeout=0` indeed.

How does each process try to insert this data? Many promises run in parallel / at the same time?

**Comment by JonathanCallewaert** [comment]
Yes you're right. A lot of promises in parallel/at the same time. There is a `Promise.all` of 250 items for each child process. It are all create ore update operations. The only reason is that I should get this migration as fast as possible as our platform is temporary going down. Maybe I can use the transaction API to solve my problem? I will test later. 

What's the max pool timeout I can set? Although, it would be nice if the`pool_timeout` would work as expected.

**Comment by janpio** [comment]
We found a limitation in the current version that even with `pool_timeout=0` (which should disable the timeout) runs into a timeout at ~30 seconds after you start. Does this match your observations when you see the error?

**Comment by JonathanCallewaert** [comment]
Yes that's possible!

I also notices something else yesterday. The postgres DB is an RDS on AWS.
I got this issue when writing to that RDS on my local machine through a VPN (my internet speed is +- 200mbps fyi).
When I run a script on an ECS instance on AWS, which doesn't need VPN to access the RDS but just can access it, I do not have the problem anymore! When I test this script on my local machine, and do a lot of writes to a local db, I also don't have this problem. So apparently I only have this issue when writing to a remote DB. 

I hope this info can help you to discover the bug.

**Comment by janpio** [comment]
Some possible explanation: Writing to a remote database would increase the roundtrip per query, leading to longer query runtime, leading to longer queue time, possibly triggering the timeout. When connection locally, you do not pass that threshold. Still a bit confusing...

**Comment by RoyBkker** [comment]
I have the the same issue on Google Cloud Platform, running a Cloud Run instance (with one minimum instance for autoscaling), with PubSub push subscription.  We are having Prisma 2.7.1. in production.

**Comment by janpio** [comment]
> We are having Prisma 2.7.1. in production.

Please try to use a more recent version, we have made quite a few changes to the timeout handling that should improve things.

**Comment by pantharshit00** [comment]
Hi here,

We did some fixes in the dev channel. I can't reproduce this now with `?pool_timeout=0&connection_limit=18` (Increase connection limit according to your database limits). 

Please try the dev channel and let me know if this is fixed for you.
```
npm uninstall @prisma/cli @prisma/client
npm install prisma@dev @prisma/client@dev
```

(Note: We are going to change package name for cli(from `@prisma/cli` to `prisma`) in next release so please make sure you have the right thing).

**Comment by mjknight50** [comment]
I just had the connection pool issue in production:
`Error: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 3, amount of connections currently in use: 0.`

 "@prisma/cli": "^2.15.0",
"@prisma/client": "^2.15.0",

**Comment by mjknight50** [comment]
I just had this error in production:
`PrismaClientInitializationError2 [PrismaClientInitializationError]: Timed out fetching a new connection from the pool. Please consider reducing the number of requests or increasing the `connection_limit` parameter (https://www.prisma.io/docs/concepts/components/prisma-client/connection-management#connection-pool). Current limit: 3, amount of connections currently in use: 0.`

After chasing it around a bit, I realized that it is actually that I had an incorrect connection string and had nothing to do with connection_limit at all.

Perhaps some better error responses would be helpful?

**Comment by matthewmueller** [comment]
@mjknight50 so it looks like you have a current connection limit of 3, you can comfortably have 3 concurrent prisma queries sent at once. 

Another possibility: if you're running the Prisma Client inside a node http server, each handler runs concurrently, so you can have 3 simultaneous visitors each sending 1 query. 

If you're above this, these requests will get queued and processed in FIFO order. If the requests spend too long in the queue, they will start to timeout. 

This is the problem you're running into and the solution is to either reducing the number of requests or increase the connection_limit parameter.

**Comment by matthewmueller** [comment]
FYI, we're actively working on #4998 which should alleviate one of the main causes of this problem. With #4998, you'll be able to insert many records at once on a single connection.

**Comment by mjknight50** [comment]
@matthewmueller You are right, that did fix it.  Thank you.
For what it's worth, I was using Prisma in an EC2 task as part of a node job, writing to mySQL on AWS.  It wasn't long running, nor have I seen this in previous production cases.    In any event, it now works.  Thank you.

**Comment by pantharshit00** [comment]
Try the pool options with the 2.16 release now `?pool_timeout=0&connection_limit=18`

**Comment by JonathanCallewaert** [comment]
Still getting those errors sometimes on 2.16. Event with pool_timeout=0. Happens when doing 20 parallel read queries on the same table

**Comment by pantharshit00** [comment]
@JonathanCallewaert Where is your database hosted?

**Comment by JonathanCallewaert** [comment]
It's a Postgress db hosted on AWS (Amazon Aurora)

**Comment by shwethashyam1** [comment]
I think this is related to the fixed connection limit, for e.g. I have a two core CPU and my connection limit is fixed at 5.
https://github.com/prisma/quaint/blob/79130113115d4a5f016e7d447b13a81aa778d59e/src/pooled.rs#L187

**Comment by janpio** [comment]
The error message this thread is/was about does not exist in Prisma Client and its engines any more. If you have a related problem, please open a new issue with a full description of what is happening for you. Thanks.

**Comment by nicosampler** [comment]
@janpio was the fix added to the docs?

**Comment by janpio** [comment]
Which fix are you referring to? We made the error message better and more descriptive, if you encounter any of these please open an issue and we can investigate what is causing it.

**Comment by r0b0tt** [comment]
I had the same issue using a digital ocean database. The issue was that my IP address was not whitelisted.

**Comment by C-Sinclair** [comment]
> I had the same issue using a digital ocean database. The issue was that my IP address was not whitelisted.

I had literally tried everything with this error, and yes... it turned out to be the IP whitelist for me too! :doughnut:
